# **PREDICT DIABETES**

**KNN**

K Nearest Neighbors is one of the simplest Supervised Machine Learning algorithm mostly used for classification. It classifies a data point based on how its neighbors are classified.

In [ ]:
# Import the required libraries

import numpy as np
import pandas as pd
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import train_test_split

In [ ]:
# Import Dataset

data = pd.read_csv('/content/diabetes.csv')


In [ ]:
print(len(data))

768


In [ ]:
data.shape

(768, 9)

In [ ]:
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [ ]:
data.isnull().any()

Pregnancies                 False
Glucose                     False
BloodPressure               False
SkinThickness               False
Insulin                     False
BMI                         False
DiabetesPedigreeFunction    False
Age                         False
Outcome                     False
dtype: bool

In [ ]:
data.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

In [ ]:
data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

Values of columns like 'Glucose', 'BloodPressure' cannot be accepted as zeroes because it will affect the outcome. We can replace such values with the mean of the respective column

In [ ]:
# Replace zeroes

replace_zeroes = ['Glucose', 'BloodPressure', 'SkinThickness', 'BMI', 'Insulin']

In [ ]:
for column in replace_zeroes :
  data[column] = data[column].replace(0, np.NaN)
  mean = int(data[column].mean(skipna = True))
  data[column] = data[column].replace(np.NaN, mean)

In [ ]:
data.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness    Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.00000   
mean      3.845052  121.682292      72.386719      29.108073  155.28125   
std       3.369578   30.435999      12.096642       8.791221   85.02155   
min       0.000000   44.000000      24.000000       7.000000   14.00000   
25%       1.000000   99.750000      64.000000      25.000000  121.50000   
50%       3.000000  117.000000      72.000000      29.000000  155.00000   
75%       6.000000  140.250000      80.000000      32.000000  155.00000   
max      17.000000  199.000000     122.000000      99.000000  846.00000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    32.450911                  0.471876   33.240885    0.348958  
std      6.875366                  0.331329   11.760232    0.476951  
min     18.200000                  0.078000   21.000000    0.000000  
25%     27.500000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

All the 0's have been replaced by mean

In [ ]:
# Split the dataset into train and test

X = data.iloc[:, 0:8]
Y = data.iloc[:, 8]
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, random_state = 0, test_size = 0.2)

**Feature Scaling**

In [ ]:
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.fit_transform(X_test)

Define the model using KNeighborsClassifier and fit the train data in the model.N_neighbors here is 'K' p is the power parameter to define the metric used. Which is 'Euclidean' in our case

In [ ]:
import math
math.sqrt(len(Y_test))

12.409673645990857

In here the square root of the sum of the squared differences between the coordinates of two points is a even number. To avoid ties, it is recommended to use an odd value of k

In [ ]:
# Define the model : Init KNN

classifier = KNeighborsClassifier(n_neighbors = 11, p = 2, metric = 'euclidean')

It's important to evaluate the model, Let's use confusion matrix to do that

In [ ]:
# Fit Model
classifier.fit(X_train, Y_train)

KNeighborsClassifier(metric='euclidean', n_neighbors=11)

In [ ]:
Y_predict = classifier.predict(X_test)
Y_predict

array([1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1,
       1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1,
       1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [ ]:
# Evaluate Model

cm = confusion_matrix(Y_test, Y_predict)
print (cm)

[[95 12]
 [18 29]]


In [ ]:
print (f1_score(Y_test, Y_predict))

0.6590909090909092


In [ ]:
print (accuracy_score(Y_test, Y_predict))

0.8051948051948052


We have developed a KNN model that can predict if a person will develop diabetes or not, or at the very least, if they should get checked out and have their blood sugar checked on a regular basis. The accurate score we obtained, 0.80519, was fairly close to what we obtained, so we can pretty much round that off and say that our accuracy of 80% indicates that the model fits the data fairly well.